mempool/src/pool/orphans/mod.rs

```diff
@@ -0,0 +1,267 @@
+            OutPointSourceId::Transaction(tx_id) => self.maps.by_tx_id.contains_key(&tx_id),
+            OutPointSourceId::BlockReward(_) => false,
+        })
+    }
+
+    pub fn len(&self) -> usize {
+        self.transactions.len()
+    }
+
+    /// Convert into transaction entries (in arbitrary order)
+    pub fn into_transactions(self) -> impl Iterator<Item = TxEntry> {
+        self.transactions.into_iter()
+    }
+
+    /// Insert a transaction entry
+    pub fn insert(&mut self, entry: TxEntry) -> Result<(), OrphanPoolError> {
+        let tx_id = *entry.tx_id();
+        ensure!(!self.contains(&tx_id), OrphanPoolError::Duplicate);
+
+        ensure!(
+            !entry.outpoints().any(|outpt| self.maps.by_spent_output.contains_key(outpt)),
+            OrphanPoolError::Conflict,
+        );
+
+        self.maps.insert(&entry, InternalId::new(self.len()));
+        self.transactions.push(entry);
+
+        if self.enforce_max_size(ORPHAN_POOL_SIZE_HARD_LIMIT) > 0 {
+            log::warn!("Orphan pool size hard limit hit");
+            ensure!(self.contains(&tx_id), OrphanPoolError::Full);
+        }
+
+        Ok(())
+    }
+
+    /// Insert a transaction entry and make sure the pool size does not grow too large
+    pub fn insert_and_enforce_limits(
+        &mut self,
+        entry: TxEntry,
+        cur_time: Time,
+    ) -> Result<(), OrphanPoolError> {
+        let tx_id = *entry.tx_id();
+
+        self.insert(entry)?;
+        self.enforce_limits(cur_time);
+        ensure!(self.contains(&tx_id), OrphanPoolError::Full);
+
+        Ok(())
+    }
+
+    /// Remove given transaction
+    pub fn remove(&mut self, id: Id<Transaction>) -> Option<TxEntry> {
+        self.maps.by_tx_id.get(&id).copied().map(|iid| self.remove_at(iid))
+    }
+
+    fn remove_at(&mut self, iid: InternalId) -> TxEntry {
+        let entry = self.transactions.swap_remove(iid.get());
+        self.maps.remove(&entry);
+
+        // The above swap_remove may have moved an entry from the back of the vector to the
+        // original position of the transaction we just removed. Maps have to be updated.
+        if let Some(moved_entry) = self.transactions.get(iid.get()) {
+            self.maps.remove(moved_entry);
+            self.maps.insert(moved_entry, iid);
+        }
+
+        entry
+    }
+
+    /// Enforce expiration and transaction count limits. Returns the number of erased transactions.
+    fn enforce_limits(&mut self, cur_time: Time) -> usize {
+        let n_expired = self.remove_expired(cur_time);
+        let n_evicted = self.enforce_max_size(*self.transaction_count_limit);
+        n_expired + n_evicted
+    }
+
+    fn remove_expired(&mut self, cur_time: Time) -> usize {
+        // Remove all expired txns
+        let expiry = cur_time.saturating_sub(config::DEFAULT_ORPHAN_TX_EXPIRY_INTERVAL);
+
+        let mut n_expired = 0;
+        while let Some(entry) = self.maps.by_insertion_time.first().filter(|(t, _)| *t < expiry) {
```

Comment:
I wonder whether we can use `take_while()` here... seems appropriate.

Reply:
I don't think so. The iterator would have to take a shared reference to the map but the map is also being modified in the loop body.